### ISSUE TITLE
Impact of Training's parameters on Networks performances

### ISSUE THREAD

**Issue by Hyvenos** [AlexeyAB/darknet#4745]
Hello,
I'm trying different training on a modified version of Coco, only changing the cfg file's parameters to have a view on how they impact on the networks results.
I'm using this repo.

The first thing that surprised me is the impact when changing the subdivisions. I supposed it was only a matter of performance but I trained 15 networks using batch 64/subdivision 64 and changing the other ones and all of them gave me average results. Just trying a simple yoloV3 320 batch 64/subdivision 16 gave me far better results.
I read on some places that subdivisions actually impact the training and not only in term of performance, thus I can't guess how.
In some thread it is said that it is related to this code, in network.c: 
`forward_network(net, state);
    backward_network(net, state);
    float error = get_network_cost(net);
    if(((*net.seen)/net.batch)%net.subdivisions == 0) update_network(net);`
But the condition seems right to me as net.batch is already divided by subdivision.

I tried to lower the subdivision as low as possible. Decreasing it to 8 for example gave me ~2% extra mAP. But processing the metrics (TP,FP,FN,Precision,Recall,F1) with darknet detector test and comparing its results to the ground truth in my validation test show me that the network trained with subdivision 8 is better than the same one trained with subdivision 16, but only if I set my confidence threshold to 25%. I tried to filter the predictions with confidence threshold of 50% and 80% and the one with subdivision 16 has a better f1 score. It seems that the network trained with subdivision 8 is less confident with its prediction thus I can't explain why.

I also tried to increase the input size: 320 416 608 and I need to trained a 800, but the improvements were not significant.

I need to try with modified anchors but your recommendations on the github are not encouraging and I guess the default ones are well fit for Coco? Would it make sense to change the number of anchors, let says 12 instead of 9, to cover a wider variety of bounding box?

I could increase the batch size but I can't guess how it will impact my training. My feelings about it is that the wider the batch is, the more images will be used to compute the weights, the better will be the update. Am I right?

For the learning rate decrease, you advise to use max batch_size=4000*classes. But I found that my network still learned after these iterations. So I empirically set my  maxbatche size so the network can see at least twice my train set and some other iterations when the learning rate is decreased, but I don't know if it's the right way to do.

Thanks for the help!

**Comment by AlexeyAB** [comment]
* `mini_batch = batch / subdivisions` - the higher mini_batch size, the faster training and better accuracy AP. About mini_batch and higher accuracy https://github.com/AlexeyAB/darknet/issues/4386

---

> I also tried to increase the input size: 320 416 608 and I need to trained a 800, but the improvements were not significant.

Increasing resolution **mainly** increases accuracy for small objects.

----

> I need to try with modified anchors but your recommendations on the github are not encouraging and I guess the default ones are well fit for Coco? Would it make sense to change the number of anchors, let says 12 instead of 9, to cover a wider variety of bounding box?

Yes, you can change anchors, but you should read 10 times this recommendation: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> Only if you are an expert in neural detection networks - recalculate anchors for your dataset for width and height from cfg-file: darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. But you should change indexes of anchors masks= for each [yolo]-layer, so that 1st-[yolo]-layer has anchors larger than 60x60, 2nd larger than 30x30, 3rd remaining. Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.

----

> For the learning rate decrease, you advise to use max batch_size=4000*classes. But I found that my network still learned after these iterations. 

This is only a recommendation. https://github.com/AlexeyAB/darknet#when-should-i-stop-training

> **Usually** sufficient 2000 iterations for each class(object), but not less than 4000 iterations in total. But for a more precise definition when you should stop training, use the following manual:
